fix gamepad support

tcamargo opened the pull request:
Based on libretro-mame2016.
Should fix #43.

tcamargo committed:
fix gamepad support

Based on libretro-mame2016.
Should fix #43.

tcamargo commented:
Any feedback? @twinaphex

hizzlekizzle commented:
As mentioned on Discord, we should probably change this to use RETRO_DEVICE_JOYPAD instead of the keyboard, since the way 2016 does it is... weird.

tcamargo commented:
Thx! I tried to use RETRO_DEVICE_JOYPAD without success. Ideas?

Tatsuya79 commented:
MAME stand-alone default is doing just the same: mapping direction on the analog left stick. (you can hit "delete" twice on a mapping to bring it back to default.

I can understand wanting to make it easier for users but I wonder if it's going to mess up the default mapping in other games that use the analog stick (changing that to dpad or having dpad directions mapped to several conflicting commands).

tcamargo commented:
UDLR is mapped to both dpad and left analog. The original core code also mapped directionals to dpad, but it was not working.

bigianb commented:
Given that the current state makes this core unusable on a cab with a discrete joystick (and this hopefully fixes that) ... does it have any bad side effects that would prevent merging it?

tcamargo commented:
No side effect that I am aware of.

tcamargo commented:
@bigianb as a workaround, for current core, you can map you discrete joy to left analog inputs in RA.

bigianb commented:
> @bigianb as a workaround, for current core, you can map you discrete joy to left analog inputs in RA.

Ah ... yes of course ... just tested that and it works! Thank you (somewhat embarrassing not to figure that out ;-) )

tcamargo commented:
It is also possible to workaround by using a custom cfg. You can generate one via MAME UI (TAB) and remap Px Up/Down/Left/Right.
```
<?xml version="1.0"?>
<!-- This file is autogenerated; comments and unknown tags will be stripped -->
<mameconfig version="10">
    <system name="default">
        <input>
            <port type="P1_JOYSTICK_UP">
                <newseq type="standard">
                    JOYCODE_1_BUTTON11
                </newseq>
            </port>
            <port type="P1_JOYSTICK_DOWN">
                <newseq type="standard">
                    JOYCODE_1_BUTTON12
                </newseq>
            </port>
            <port type="P1_JOYSTICK_LEFT">
                <newseq type="standard">
                    JOYCODE_1_BUTTON13
                </newseq>
            </port>
            <port type="P1_JOYSTICK_RIGHT>
                <newseq type="standard">
                    JOYCODE_1_BUTTON14
                </newseq>
            </port>
        </input>
    </system>
</mameconfig>
```
I am thinking about changing to HATxUP/DOWN/LEFT/RIGHT (input_uwp does that). Anyway, HAT is also not mapped by default. A cfg file is still needed.

tcamargo committed:
rename axes and correct L2/R2 axes mapping

tcamargo committed:
use log_cb instead of fprintf

tcamargo committed:
map L2 to Button7 and R2 to Button8

tcamargo committed:
optimize retro_joystick_device

tcamargo committed:
remap d-pad ITEM_ID_HAT<x><DIR>

tcamargo committed:
add d-pad(ITEM_ID_HAT<x><DIR>) to default mame mapping

d-pad and analog should work out of the box without an external mapping 
file

tcamargo committed:
partially revert 75db5aafa49818a9021cc08ca785377b823a483d

L2/R2 mapping to B7 and B8 is not necessary

tcamargo commented:
I finally got an understating how inputs work in mame.

* Removed the keyboard device hack. I think it was used to avoid clashes. You can only have one button per input_item_id;
* Remapped d-pad to HAT input ids;
* Change default UDLR mapping to consider HAT ids.
* Cleaned-up input_retro.cpp a bit

Therefore, d-pad and analog inputs should work out of the box without the need for an external mapping file.

tcamargo committed:
undo (correct) L2/R2 mapping

Tatsuya79 commented:
We don't have a default dpad bind for UI directions (Input (general) > User Interface > UI Up/Down/Left/Right). This should be added like for player controls.

As it is now you're losing UI directions in MAME GUI if you had something custom set there.

...but if we do it like for players with "Up or Dpad Up or JOY1 UP" I wonder if moving with the keyboard in the menu isn't going to do a ghost input (moving twice) as in RA the kb arrows are doing kb and also dpad directions together.

edit: well it should be fine as it's not doing double tap in game, sent a PR
https://github.com/libretro/mame/pull/127

tcamargo commented:
It wont double tap. Currently, it has the same binding as normal UP/DOWN/LEFT/RIGHT. So, double tap would be happening now.

1985a commented:
Finally, I want to say thank you so much. I have been using a DS controller and I was a pain to play with the joystick. I'm from the old school so, always prefer to play with DPAD.
